src/components/CallView/Grid/Grid.vue

```diff
@@ -828,6 +928,81 @@ export default {
 			return callParticipantModel.attributes.peerId === this.$store.getters.selectedVideoPeerId
 		},
 
+		isModelWithVideo(callParticipantModel) {
+			return callParticipantModel.attributes.videoAvailable
+				&& (typeof callParticipantModel.attributes.stream === 'object')
+		},
+
+		isModelWithAudio(callParticipantModel) {
+			const participant = this.$store.getters.getParticipantBySessionId(this.token, callParticipantModel.attributes.nextcloudSessionId)
+			if (!participant) {
+				return false
+			}
+			return participant?.permissions & PARTICIPANT.PERMISSIONS.PUBLISH_AUDIO
+		},
+
+		unpromoteSpeaker(model) {
+			const id = model.attributes.nextcloudSessionId
+			// remove the model from the temp promoted speakers
+			const index = this.videosTempPromoted.indexOf(model)
+			if (index === -1) {
+				return
+			}
+
+			this.videosTempPromoted.splice(index, 1)
+
+			if (this.isModelWithVideo(model)) {
+				if (!this.tilesWithVideoMask.includes(id)) {
+					this.tilesWithVideoMask.unshift(id)
+				}
+			} else {
+				if (!this.tilesWithAudioMask.includes(id)) {
+					this.tilesWithAudioMask.unshift(id)
+				}
+			}
+		},
+
+		promoteSpeaker(model) {
+			const id = model.attributes.nextcloudSessionId
+
+			// if model is already in the first page, do nothing
+			if (this.orderedVideos.slice(0, this.slots).find(video => video.attributes.nextcloudSessionId === model.attributes.nextcloudSessionId)) {
+				return
+			}
+
+			if (this.screens.includes(model.attributes.peerId)) {
+				// tiles with screenshare have a better priority position already
+				// do nothing
+				return
+			}
+
+			// add the model
+			if (!this.videosTempPromoted.includes(model)) {
+				const videoIndex = this.tilesWithVideoMask.indexOf(id)
+				if (videoIndex !== -1) {
+					this.tilesWithVideoMask.splice(videoIndex, 1)
+				}
+				const audioIndex = this.tilesWithAudioMask.indexOf(id)
+				if (audioIndex !== -1) {
+					this.tilesWithAudioMask.splice(audioIndex, 1)
+				}
+
+				this.videosTempPromoted.unshift(model)
+			}
+		},
+
+		getOrderedTiles(tilesMap, orderMask) {
+			// Get the ordered tiles
+			const orderedTiles = orderMask
+				.filter(id => tilesMap.has(id))
+				.map(id => tilesMap.get(id))
+
+			// Now the rest of allTiles that are not in the orderedTiles
+			const rest = Array.from(tilesMap.values())
+				.filter(tile => !orderedTiles.includes(tile))
```

Comment:
Let's try to do it in one iteration?
```suggestion
			const { orderedTiles, rest } = orderMask.reduce((acc, id) => {
				if (tilesMap.has(id)) {
					acc.orderedTiles.push(tilesMap.get(id))
				} else {
					acc.rest.push(tilesMap.get(id))
				}
				return acc
			}, { orderedTiles: [], rest: [] })
```

Reply:
You're right, that wouldn't work here